### PR TITLE
Inital pull request to decoder

### PR DESCRIPTION
**Pull request by generall**
Hi, this is just a initial pull-request to signal the start of work.

I created some basic interfaces, here is the main thought:
Each part of the decoder that depends on Vocabulary should stay in Model class, including BeamSearch, sample scheduler, `_output_projection_layer`.

Decoder should implement same functionality as `_prepare_output_projections` implements now with some minor changes.

### CODE REVIEW

**Comment by epwalsh** [comment]
Thanks @generall! My one comment is that I think the `.forward()` should implement the entire forward loop, not just a single decoding step. What do you think?

Maybe the base decoder could implement the entire `.forward()` method, where it would call `.forward_step()` or something (which would be implemented by the subclasses). What do you think?

**Comment by generall** [comment]
Hi, I can see several reasons why `_forward_loop` should not be implemented in decoder:

1. `_forward_loop` does not contain all decoder code. It also presence in `take_step` function which is used inside BeamSearch
2. To implement `_forward_loop` functionality we need to have `_end_index` and `num_classes`. This values depend on `vocab`, and AFAIK we can't directly pass vocabulary to non-model class on `__init__`. In this case, we will have to invent additional initialization inside Seq2Seq2.
3. As I understand, it is good to have components as simple as possible. Since we don't want to re-implement BeamSearch and other tricks for better training in every decoder class.

Of course it is desirable to decompose BeamSearch and other training code out from Seq2Seq, but I think that 
this code should not be implemented inside Registrable component.

**Comment by epwalsh** [comment]
Gotcha. I think that's fair considering decoders could be pretty general, and we don't want to restrict ourselves to basic RNN decoders.

File: allennlp/modules/seq2seq_decoders/seq2seq2_decoder.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+class Seq2SeqDecoder(torch.nn.Module, Registrable):
+    # pylint: disable=abstract-method
+    """
+    A ``Seq2SeqDecoder`` is a ``Module`` that takes as input a result of ``Seq2SeqEncoder`` and returns a
+    new sequence of vectors.
+
+    Decoder should implement function to perform a single decoding step.
+    """
+
+    def __init__(self, output_dim):
```
Comment (epwalsh) [inline allennlp/modules/seq2seq_decoders/seq2seq2_decoder.py]: I don't think `output_dim` should be passed to `__init__`. Usually the decoder will output a `(batch_size, sequence_length)`, so `output_dim` would be `sequence_length`, but `sequence_length` is variable.

File: allennlp/modules/seq2seq_decoders/seq2seq2_decoder.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        super(Seq2SeqDecoder, self).__init__()
+        self._output_dim = output_dim
+
+    def get_output_dim(self) -> int:
```
Comment (epwalsh) [inline allennlp/modules/seq2seq_decoders/seq2seq2_decoder.py]: Related to my other comment, we probably don't need this

**Review by epwalsh — CHANGES_REQUESTED** [review]

File: allennlp/modules/seq2seq_decoders/seq2seq2_decoder.py (hunk 1 above)
Comment (generall) [reply allennlp/modules/seq2seq_decoders/seq2seq2_decoder.py]: Since we agreed, that decoder should be separated from Vocabulary and all Embeddings, we can't decode actual output token id inside decoder. We can only output predicted target embedding which should have dimensionality `output_dim` (or as I renamed it later `decoding_dim`). It's also important to return embedding for `BeamSearch`, because it may chose different indexes based on random search policy.

**Comment by generall** [comment]
My idea is to make a decoder in such a way that it can be implemented in this part.
decoder.

We can also decompose all decoder-related functions out from SimpleSeq2Seq, but let's do it as separate part of this issue and focus first on decomposing of NN-related part of decoder.

File: allennlp/modules/seq2seq_decoders/seq2seq2_decoder.py (hunk 2 above)
Comment (generall) [reply allennlp/modules/seq2seq_decoders/seq2seq2_decoder.py]: Same here

**Comment by epwalsh** [comment]
@generall thanks for this and sorry I've been slow to respond. I'll take a look first thing tomorrow

File: allennlp/modules/seq2seq_decoders/seq2seq2_decoder.py (hunk 1 above)
Comment (epwalsh) [reply allennlp/modules/seq2seq_decoders/seq2seq2_decoder.py]: IMHO a `Seq2SeqDecoder` should be in charge of outputting an entire sequence, not just a single hidden state or prediction from one timestep, otherwise it should probably be named differently. I'm also not convinced that this abstraction for a "seq2step" is general enough. As is, it wouldn't exactly work for CopyNet: https://github.com/allenai/allennlp/blob/master/allennlp/models/encoder_decoders/copynet_seq2seq.py

Comment (generall) [reply allennlp/modules/seq2seq_decoders/seq2seq2_decoder.py]: Ok, good point on CopyNet. I will think about how to incorporate it into new decoder implementation. Agreed that current `Seq2SeqDecoder` should be called differently

**Comment by generall** [comment]
Hi, I have decomposed decoder from seq2seq model. Right now it consists of:

* base class `SeqDecoder`
* simple decoder implementation `SimpleSeqDecoder`
* decoder cell, which contains only decoder NN layers `SimpleDecoderCell` with base class `DecoderCell`

I had some issues during my attempts to use `SeqDecoder` in copynet. I noticed that it has it's own dataset reader and I wonder if it is possible to somehow move it's preprocessing inside copynet. In this case it will be possible to implement copynet decomposed `SeqDecoder`.

**Comment by generall** [comment]
Hi again @epwalsh , would you have a time to review my changes?

**Comment by epwalsh** [comment]
Hey @generall, sorry about the late response, this completely fell off of my radar. Honestly I don't think I'll have much time to help out with this anytime soon, but it looks like you've got a pretty good start, so I'd suggest continuing this on your branch, and making a WIP (work in progress) pull request to AllenNLP (the main repo). I'd be happy to offer suggestions before you make an official PR if you'd like.
